Adding lazy relation -> data.frame conversion for R client

hannes opened the pull request:
We add functionality to convert a R DuckDB relation object to a data.frame without actually computing the query result using lazy ALTREP vectors. This works great except for list/struct/blob columns. There is hope R will add list ALTREP support soon-ish, which should remove the limitation.

hannes committed:
Relational tweaks

krlmlr committed:
Unexport

krlmlr committed:
Export again

hannes committed:
resolving todo regarding attribute setting

hannes committed:
Merge branch 'master' of github.com:duckdb/duckdb into relaltrepdf-2

krlmlr committed:
Add methods

hannes committed:
added df_is_materialized function to check if a lazy df was already materialized

hannes committed:
Merge branch 'relaltrepdf-2' of github.com:hannes/duckdb into relaltrepdf-2

krlmlr commented on tools/rpkg/src/reltoaltrep.cpp:
```diff
@@ -0,0 +1,259 @@
+
+	AltrepRownamesWrapper(shared_ptr<AltrepRelationWrapper> rel_p) : rel(rel_p) {
+		rowlen_data[0] = NA_INTEGER;
+		rowlen_data[1] = -42;
```
Do we need this?

```suggestion
```

krlmlr commented on tools/rpkg/src/reltoaltrep.cpp:
```diff
@@ -0,0 +1,259 @@
+}
+
+R_xlen_t RelToAltrep::RownamesLength(SEXP x) {
+	return 2;
```
Magic number?

krlmlr commented on tools/rpkg/src/reltoaltrep.cpp:
```diff
@@ -0,0 +1,259 @@
+	case LogicalTypeId::UUID:
+		return RelToAltrep::string_class;
+	default:
+		cpp11::stop("rel_to_altrepUnknown column type for altrep: %s", type.ToString().c_str());
```
```suggestion
		cpp11::stop("rel_to_altrep: Unknown column type for altrep: %s", type.ToString().c_str());
```

krlmlr commented on tools/rpkg/src/reltoaltrep.cpp:
```diff
@@ -0,0 +1,259 @@
+	for (size_t col_idx = 0; col_idx < ncols; col_idx++) {
+		cpp11::external_pointer<AltrepVectorWrapper> ptr(new AltrepVectorWrapper(relation_wrapper, col_idx));
+		auto column_type = rel->Columns()[col_idx].Type();
+		auto vector_sexp = r_protector.Protect(R_new_altrep(LogicalTypeToAltrepType(column_type), ptr, R_NilValue));
```
AFAICT protection not needed here?

```suggestion
		auto vector_sexp = R_new_altrep(LogicalTypeToAltrepType(column_type), ptr, R_NilValue);
```

hannes replied:
we now need it

krlmlr reviewed (approved):
🎉

hannes committed:
updates after review

hannes committed:
found a somewhat serious issue where we re-assign the r types for every duckdb vector we transform

hannes committed:
fix special types for altrep lazy df

hannes committed:
convenience rename

hannes committed:
message when materializing

hannes committed:
once per materialize

hannes committed:
making materialization message optional

hannes committed:
changing flag to be consistent

krlmlr committed:
Avoid bogus return

hannes commented:
@krlmlr do we need to remove the examples for the un-exported methods? R CMD check does not seem to like them.

krlmlr commented:
I think so, yes.

hannes committed:
make cleanMerge branch 'master' of github.com:duckdb/duckdb into
relaltrepdf-2 # Please enter a commit message to explain why this merge is
necessary, # especially if it merges an updated upstream into a topic
branch. # # Lines starting with '#' will be ignored, and an empty message
aborts # the commit.

hannes commented:
I'm also wondering why those things have to be private in the first place. Maybe its better to mark them experimental.

hannes committed:
keep exporting rel stuff after all

krlmlr commented:
We agreed that keeping the functions private will make it easier for us later to use the new relational package. What is the advantage of keeping them exported?

krlmlr reviewed (approved):
I don't have a strong opinion regarding reexport, I see there is now a message that explicitly alerts users. Would you like a beautiful orange badge, like in https://dplyr.tidyverse.org/reference/rows.html?

hannes committed:
@noRd ftw

hannes commented:
> Would you like a beautiful orange badge, like in https://dplyr.tidyverse.org/reference/rows.html?

Since this pulls in the lifecycle package I am adding `@noRd` as discussed

hannes committed:
Merge branch 'master' of github.com:duckdb/duckdb into relaltrepdf-2

hannes committed:
hide rel object in data2 of row.names altrep
